### PR TITLE
Release notes for Percona Server 5.7.11-4

### PR DESCRIPTION
**Pull request by hrvojem**

### CODE REVIEW

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst [hunk 1]
```diff
@@ -18,17 +18,6 @@ To avoid this |Percona Server| has implemented priority refill for the buffer po
 
 Even the above implementation does not fully resolve the buffer pool free list mutex contention, as the mutex is still being acquired needlessly whenever the buffer pool free list is empty. This is addressed by delegating all the LRU flushes to the to the LRU manager thread, never attempting to evict a page or perform a LRU single page flush by a query thread, and introducing a backoff algorithm to reduce buffer pool free list mutex pressure on empty buffer pool free lists. This is controlled through a new system variable :variable:`innodb_empty_free_list_algorithm`.
  
-.. variable:: innodb_empty_free_list_algorithm
```
Comment (laurynas-biveinis) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst:21]: Why is this moved?

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst [hunk 2]
```diff
@@ -46,6 +35,52 @@ The following |InnoDB| metrics are no longer accounted, as their semantics do no
 
 The need for |InnoDB| recovery thread writer threads is also removed, consequently all associated code is deleted.
 
+.. _parallel_doublewrite_buffer:
+
+Parallel doublewrite buffer
+===========================
+
+The existing doublewrite buffer is shared between all the buffer pool instances and all the flusher threads. PMP shows that it's becoming a point of contention with flusher threads waiting for the running batch to end so that they can post a page to flush. This limits the effect of extra cleaner threads. Moreover, single page flushes wait for the above and also contend on the doublewrite mutex.
```
Comment (laurynas-biveinis) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: The legacy doublewrite buffer is shared between all the buffer pool instances and all the flusher threads. It collects all the page write requests into a single buffer, and, when the buffer fills, writes it out to disk twice, blocking any new write requests until the writes complete. This becomes a bottleneck with increased flusher parallelism, limiting the effect extra cleaner threads. In addition, single page flushes, if they are performed, are subject to above and also contend on the doublewrite mutex.

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst [hunk 3]
```diff
@@ -46,6 +35,52 @@ The following |InnoDB| metrics are no longer accounted, as their semantics do no
...
+
+The location of the doublewrite file is governed by a new :variable:`innodb_parallel_doublewrite_path` global, read-only system variable. It defaults to :file:`xb_doublewrite` in the data directory. The variable accepts both absolute and relative paths. In the latter case they are treated as relative to the data directory. The doublewrite file is not a tablespace from |InnoDB| internals point of view.
+
+The legacy |InnoDB| doublewrite buffer in the system tablespace continues to address doublewrite needs of single page flushes, and they are free to use the whole of that buffer (128 pages by default) instead of the last eight pages as currently used. Note that single page flushes will not happen in |Percona Server| unless :variable:`innodb_empty_free_list_algorithm` is set to ``legacy`` value.
+
+The existing transaction system header in the system tablespace and the existing doublewrite buffer are not touched in any way. Thus perfect cross-grade compatibility is ensured on clean shutdowns.
```
Comment (laurynas-biveinis) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: The existing system tablespace is not touched in any way for this feature implementation, ensuring that cleanly-shutdown instances may be freely moved between different server flavors.

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst [hunk 4]
```diff
@@ -46,6 +35,52 @@ The following |InnoDB| metrics are no longer accounted, as their semantics do no
...
+
+Interaction with :variable:`innodb_flush_method`
+------------------------------------------------
+
+Regardless of its setting, the parallel doublewrite file is opened with ``O_DIRECT`` flag to remove OS caching. Then its access is further governed by the :variable:`innodb_flush_method` setting: if it's set to ``O_DSYNC``, the parallel doublewrite is opened with ``O_SYNC`` flag too. Further, if it's one of ``O_DSYNC``, ``O_DIRECT_NO_FSYNC``, or ``ALL_O_DIRECT``, then the doublewrite file is not flushed after a batch of writes to it is completed.
```
Comment (laurynas-biveinis) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: This text has to be folded into the innodb_flush_method-describing page too.

Comment (laurynas-biveinis) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: s/its/innodb_flush_method, then its access is further governed by the exact value set: if it's O_DSYNC ...

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst [hunk 5]
```diff
@@ -46,6 +35,52 @@ The following |InnoDB| metrics are no longer accounted, as their semantics do no
...
+.. variable:: innodb_empty_free_list_algorithm
+
+   :cli: Yes
+   :conf: Yes
+   :scope: Global
+   :dyn: Yes
+   :values: legacy, backoff
+   :default: backoff
+
+When ``legacy`` option is set, server will used the upstream algorithm and when the ``backoff`` is selected, |Percona| implementation will be used.
```
Comment (laurynas-biveinis) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: If there's reason to move it here, then the description is without context

File: doc/source/release-notes/Percona-Server-5.7.11-4.rst [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ :ref:`tokudb_background_analyze_table` feature is now enabled by default (:variable:`tokudb_analyze_in_background` is set to ``ON`` by default). Variable :variable:`tokudb_auto_analyze` default value has been changed from ``0`` to ``30``. (:tokubug:`935`) 
+
+ :ref:`log_warning_suppress` feature has been removed from |Percona Server| 5.7 because |MySQL| 5.7.11 has implemented a new system variable, `log_statements_unsafe_for_binlog <https://dev.mysql.com/doc/refman/5.7/en/replication-options-binary-log.html#sysvar_log_statements_unsafe_for_binlog>`_, which implements the same effect.
+
+Bugs Fixed
+==========
+
+ If ``pid-file`` option wasn't specified with the full path, *Ubuntu*/*Debian* ``sysvinit`` script wouldn't notice the server is actually running and it will timeout or in some cases even hang. Bug fixed :bug:`1549333`.
+
+ Buffer pool may fail to remove dirty pages for a particular tablespace from the flush list which could lead to debug build crash. Bug fixed :bug:`1552673`.
```
Comment (laurynas-biveinis) [inline doc/source/release-notes/Percona-Server-5.7.11-4.rst]: Buffer pool may fail to remove dirty pages for a particular tablesspace from the flush list, as requested by, for example, DROP TABLE or TRUNCATE TABLE commands. This could lead to a crash.

File: doc/source/upstream-bug-fixes.rst [hunk 7]
```diff
@@ -5,21 +5,39 @@
 =============================================================
 
 +-------------------------------------------------------------------------------------------------------------+
+|:Upstream bug: :mysqlbug:`80607` - main.log_tables-big unstable on loaded hosts                              |
+|:Launchpad bug: :bug:`1554043`                                                                               |
+|:Upstream state: Verified (checked on 2016-03-11)                                                            |
+|:Fix Released: :rn:`5.7.11-4`                                                                                |
+|:Upstream fix: N/A                                                                                           |
++-------------------------------------------------------------------------------------------------------------+
+|:Upstream bug: :mysqlbug:`80606` - Audit Log Plugin worker thread may crash on write call writing fewer ...  |
```
Comment (laurynas-biveinis) [inline doc/source/upstream-bug-fixes.rst]: The upstream bug title is "my_write, my_pwrite no longer safe to call from THD-less server utility threads"

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst (hunk 1 above)
Comment (hrvojem) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst:21]: If there are more variable descriptions they go into System Variables section.

**Comment by hrvojem** [comment]
Fixed/pushed

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst [hunk 8]
```diff
@@ -46,6 +46,38 @@ The following |InnoDB| metrics are no longer accounted, as their semantics do no
 
 The need for |InnoDB| recovery thread writer threads is also removed, consequently all associated code is deleted.
 
+.. _parallel_doublewrite_buffer:
+
+Parallel doublewrite buffer
+===========================
+
+The legacy doublewrite buffer is shared between all the buffer pool instances and all the flusher threads. It collects all the page write requests into a single buffer, and, when the buffer fills, writes it out to disk twice, blocking any new write requests until the writes complete. This becomes a bottleneck with increased flusher parallelism, limiting the effect extra cleaner threads. In addition, single page flushes, if they are performed, are subject to above and also contend on the doublewrite mutex.
```
Comment (laurynas-biveinis) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: ... limiting the effect OF extra cleaner threads

File: doc/source/known_issues.rst [hunk 9]
```diff
@@ -0,0 +1,13 @@
+.. _known_issues:
```
Comment (laurynas-biveinis) [inline doc/source/known_issues.rst:1]: Where is this page linked from besides index?

File: doc/source/scalability/innodb_io.rst [hunk 10]
```diff
@@ -65,6 +65,10 @@ The following values are allowed:
   * ``ALL_O_DIRECT``: 
     use O_DIRECT to open both data and log files, and use ``fsync()`` to flush the data files but not the log files. This option is recommended when |InnoDB| log files are big (more than 8GB), otherwise there might be even a performance degradation. **Note**: When using this option on *ext4* filesystem variable `innodb_log_write_ahead_size <https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_log_write_ahead_size>`_ should be set to 4096 (default log-block-size in *ext4*) in order to avoid the ``unaligned AIO/DIO`` warnings.
 
+Interaction with :ref:`parallel_doublewrite_buffer`
+---------------------------------------------------
+
+Regardless of :variable:`innodb_flush_method setting`, the parallel doublewrite file is opened with ``O_DIRECT`` flag to remove OS caching, then its access is further governed by the exact value set: if it's set to ``O_DSYNC``, the parallel doublewrite is opened with ``O_SYNC`` flag too. Further, if it's one of ``O_DSYNC``, ``O_DIRECT_NO_FSYNC``, or ``ALL_O_DIRECT``, then the doublewrite file is not flushed after a batch of writes to it is completed.
```
Comment (laurynas-biveinis) [inline doc/source/scalability/innodb_io.rst]: O_DIRECT bit may be removed and the rest inlined into the value descriptions above

**Comment by hrvojem** [comment]
Fixed/pushed

File: doc/source/scalability/innodb_io.rst [hunk 11]
```diff
@@ -54,17 +54,16 @@ The following values are allowed:
     use ``fsync()`` to flush both the data and log files.
 
   * ``O_SYNC``: 
-    use O_SYNC to open and flush the log files; use ``fsync()`` to flush the data files.
+    use O_SYNC to open and flush the log files; use ``fsync()`` to flush the data files. When this value is set :ref:`parallel_doublewrite_buffer` file is opened with ``O_SYNC`` flag.
```
Comment (laurynas-biveinis) [inline doc/source/scalability/innodb_io.rst]: use O_SYNC to open and flush the log and parallel doublewrite files; use fsync() to flush the data files. Do not use fsync() to flush the parallel doublewrite file.

File: doc/source/scalability/innodb_io.rst [hunk 12]
```diff
@@ -54,17 +54,16 @@ The following values are allowed:
     use ``fsync()`` to flush both the data and log files.
 
   * ``O_SYNC``: 
-    use O_SYNC to open and flush the log files; use ``fsync()`` to flush the data files.
+    use O_SYNC to open and flush the log files; use ``fsync()`` to flush the data files. When this value is set :ref:`parallel_doublewrite_buffer` file is opened with ``O_SYNC`` flag. 
 
   * ``O_DIRECT``: 
-    use O_DIRECT to open the data files and fsync() system call to flush both the data and log files.
+    use O_DIRECT to open the data files and fsync() system call to flush both the data and log files.
```
Comment (laurynas-biveinis) [inline doc/source/scalability/innodb_io.rst]: and fsync system call to flush data, log, and parallel doublewrite files.

File: doc/source/scalability/innodb_io.rst [hunk 13]
```diff
@@ -54,17 +54,16 @@ The following values are allowed:
...
-    use O_SYNC to open and flush the log files; use ``fsync()`` to flush the data files.
+    use O_SYNC to open and flush the log files; use ``fsync()`` to flush the data files. When this value is set :ref:`parallel_doublewrite_buffer` file is opened with ``O_SYNC`` flag. 
 
   * ``O_DIRECT``: 
-    use O_DIRECT to open the data files and fsync() system call to flush both the data and log files.
+    use O_DIRECT to open the data files and fsync() system call to flush both the data and log files. 
 
   * ``O_DIRECT_NO_FSYNC``:
-    use O_DIRECT to open the data files but don't use ``fsync()`` system call to flush both the data and log files. This option isn't suitable for *XFS* file system.
+    use O_DIRECT to open the data files but don't use ``fsync()`` system call to flush both the data and log files. When this value is set, the :ref:`parallel_doublewrite_buffer` file is not flushed after a batch of writes to it is completed. This option isn't suitable for *XFS* file system.
```
Comment (laurynas-biveinis) [inline doc/source/scalability/innodb_io.rst]: use O_DIRECT to open the data files, but don't use fsync() system call to flush data, log, and parallel doublewrite files.

File: doc/source/scalability/innodb_io.rst [hunk 14]
```diff
@@ -54,17 +54,16 @@ The following values are allowed:
     use ``fsync()`` to flush both the data and log files.
```
Comment (laurynas-biveinis) [inline doc/source/scalability/innodb_io.rst]: to flush data, log, and parallel doublewrite files.

File: doc/source/scalability/innodb_io.rst [hunk 15]
```diff
@@ -54,17 +54,16 @@ The following values are allowed:
...
-
+    use O_DIRECT to open both data and log files, and use ``fsync()`` to flush the data files but not the log files. This option is recommended when |InnoDB| log files are big (more than 8GB), otherwise there might be even a performance degradation. **Note**: When using this option on *ext4* filesystem variable `innodb_log_write_ahead_size <https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_log_write_ahead_size>`_ should be set to 4096 (default log-block-size in *ext4*) in order to avoid the ``unaligned AIO/DIO`` warnings. When this value is set, the :ref:`parallel_doublewrite_buffer` file is not flushed after a batch of writes to it is completed.
```
Comment (laurynas-biveinis) [inline doc/source/scalability/innodb_io.rst]: but not the log or parallel doublewrite files.

**Comment by hrvojem** [comment]
Implemented comments

File: doc/source/scalability/innodb_io.rst [hunk 16]
```diff
@@ -51,20 +51,19 @@ This is an existing |MySQL| 5.7 system variable that has a new allowed value ``A
...
-
+    use ``O_DIRECT`` to open both data and log files, and use ``fsync()`` to flush the data files but not the log or parallel doublewrite files. This option is recommended when |InnoDB| log files are big (more than 8GB), otherwise there might be even a performance degradation. **Note**: When using this option on *ext4* filesystem variable `innodb_log_write_ahead_size <https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_log_write_ahead_size>`_ should be set to 4096 (default log-block-size in *ext4*) in order to avoid the ``unaligned AIO/DIO`` warnings. When this value is set, the :ref:`parallel_doublewrite_buffer` file is not flushed after a batch of writes to it is completed.
```
Comment (laurynas-biveinis) [inline doc/source/scalability/innodb_io.rst]: The last sentence is redundant now

**Comment by laurynas-biveinis** [comment]
LGTM with that last sentence removed
